recipe-server/client/control/state/app/experimenter/reducers.js

```diff
@@ -0,0 +1,37 @@
+import { combineReducers } from 'redux';
+import { fromJS, Map, List } from 'immutable';
+
+import {
+  EXPERIMENTS_RECEIVE,
+  EXPERIMENT_ACCEPTED,
+  EXPERIMENT_ALREADY_ACCEPTED,
+  EXPERIMENT_REJECTED,
+  EXPERIMENT_SELECT_REJECTED,
+} from 'control/state/action-types';
+
+function filterExperiments(state, slug) {
+  const experiments = state.get('experiments', new List());
+  const filteredExperiments = experiments.filter(experiment => experiment.get('experiment_slug') !== slug);
+  return state.set('experiments', filteredExperiments);
+}
+
+function items(state = new Map(), action) {
+  switch (action.type) {
+    case EXPERIMENTS_RECEIVE:
+      return state.set('experiments', fromJS(action.experiments));
+    case EXPERIMENT_ACCEPTED:
+      return filterExperiments(state, action.experiment.experiment_slug);
+    case EXPERIMENT_ALREADY_ACCEPTED:
+      return filterExperiments(state, action.experiment.experiment_slug);
+    case EXPERIMENT_SELECT_REJECTED:
+      return state.set('rejected_experiment', fromJS(action.experiment));
+    case EXPERIMENT_REJECTED:
+      return filterExperiments(state, action.experiment.experiment_slug);
+    default:
+      return state;
+  }
+}
+
+export default combineReducers({
+  items,
```

Comment:
I think I tried that originally but then it didn't make sense to have state.app.experiments.experiments rather than state.app.experiments.items.experiments.  It seemed like to be consistent with the rest of the application, I should name space my whole state into something called 'items', which I think @rehandalal recommended?

Reply:
Right, so `rejected_experiments` shouldn't be nested under items. `rejected_experiments` should be it's own reducer. Something like:

```js
function items(state = new List(), action) {
  // handles RECEIVE, ACCEPTED, ALREADY_ACCEPTED, and REJECTED
}

function rejected(state = null, action) {
  switch (action.type) {
    case EXPERIMENT_SELECT_REJECTED: {
      return fromJS(action.experiment);
    }
    default: {
      return state;
    }
}

export default combineReducers({
  items,
  rejected,
});
```

Then you'd have `state.app.experiments.items` and `state.app.experiments.rejected`. To me this makes more sense than nesting `rejected` under `items`.

This is a bigger change, but most of the time that we have a collection and then a selection into that collection, like here, we use a `Map` for the items, and instead of storing the entire item in `rejected`, it would be `rejectedId`, and just store the key into that map for the rejected item. I'm not sure if this would work well with Experimenter's data model though.

Reply:
Ya makes sense, sounds good! That helps keep things clearer.